Dispatcher readiness e2e test

lberk opened the pull request:
e2e tests for #433 

## Proposed Changes
- e2e test to create sub, scale dispatcher, add second sub and verify readiness after dispatcher scales

lberk committed:
Parameterize Partitions created via MustCreateTopic

lberk committed:
Add kafkaChannel creation/get obj helper functions

lberk committed:
Add subscription scale readiness test

lberk commented:
/hold

lberk committed:
Add explicit t.Run statement

matzew commented on test/e2e/helpers/kafka_helper.go:
```diff
@@ -215,7 +215,7 @@ func MustPublishKafkaMessageViaBinding(client *testlib.Client, selector map[stri
 }
 
-func MustCreateTopic(client *testlib.Client, clusterName string, clusterNamespace string, topicName string) {
+func MustCreateTopic(client *testlib.Client, clusterName, clusterNamespace, topicName string, partitions int) {
```
nice enhancement to drive the partitions !

lberk committed:
Move test to helper function, use channelTestRunner

lberk committed:
Change from subscriptionv1beta1 to subscriptionv1 (abstraction prep)

lberk committed:
Support subscription v1beta1 and v1

lberk committed:
Fix copyright header

lberk committed:
Change channel creation depnding on typemeta

Remove subscription v1beta1 testing, having multiple tests run at once
will change the dispatcher replication in parallel

lberk committed:
Drop unneeded parameter

slinkydeveloper commented on test/lib/creation.go:
```diff
@@ -25,11 +25,70 @@ import (
+	}
+}
+
+func GetKafkaChannelV1Alpha1OrFail(c *testlib.Client, kafkaChannel string) *channelsv1alpha1.KafkaChannel {
```
Do we need to tes v1alpha1? We're going to remove this code in a week when the next release is out :)

lberk replied:
I didn't think so, but I also assumed if I **didn't** add it, somebody would say I **should** add it. so, :shrug:

slinkydeveloper replied:
For me you should remove them :smile:

lberk replied:
ok, removed them

slinkydeveloper commented on test/e2e/channel_subscription_ready_test.go:
```diff
@@ -0,0 +1,29 @@
+// +build e2e
+/*
```
add a blank line here

slinkydeveloper replied:
> To distinguish build constraints from package documentation, a series of build constraints must be followed by a blank line. 

From https://golang.org/cmd/go/#hdr-Build_constraints

lberk committed:
Add a blank line

lberk committed:
Fix linter complaints

lberk committed:
Add build constraint tag to mark testing consolidated channel only

lberk committed:
run hack/update-codegen.sh

lberk commented:
@devguyio any tips on why the second subscription isn't becoming ready? from the error logs:
```
Object that did not become ready {Subscription messaging.knative.dev/v1beta1}-kafka-sub-1 when dumping error state: &{TypeMeta:{Kind:Subscription APIVersion:messaging.knative.dev/v1beta1} ObjectMeta:{Name:kafka-sub-1 GenerateName: Namespace:test-channel-subscription-scale-ready-v1-kafka-channel-mesgd4h6 SelfLink:/apis/messaging.knative.dev/v1beta1/namespaces/test-channel-subscription-scale-ready-v1-kafka-channel-mesgd4h6/subscriptions/kafka-sub-1 UID:e63788d8-4e91-417c-938c-62cbf15fa61c ResourceVersion:24174 Generation:1 CreationTimestamp:2021-03-25 15:36:47 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[messaging.knative.dev/creator:prow-job@knative-tests.iam.gserviceaccount.com messaging.knative.dev/lastModifier:prow-job@knative-tests.iam.gserviceaccount.com] OwnerReferences:[] Finalizers:[subscriptions.messaging.knative.dev] ClusterName: ManagedFields:[{Manager:e2e.test Operation:Update APIVersion:messaging.knative.dev/v1 Time:2021-03-25 15:36:47 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:spec":{".":{},"f:channel":{".":{},"f:apiVersion":{},"f:kind":{},"f:name":{}},"f:subscriber":{".":{},"f:ref":{".":{},"f:apiVersion":{},"f:kind":{},"f:name":{}}}}}} {Manager:controller Operation:Update APIVersion:messaging.knative.dev/v1 Time:2021-03-25 15:36:48 +0000 UTC FieldsType:FieldsV1 FieldsV1:{"f:metadata":{"f:finalizers":{".":{},"v:\"subscriptions.messaging.knative.dev\"":{}}},"f:status":{".":{},"f:conditions":{},"f:observedGeneration":{},"f:physicalSubscription":{".":{},"f:subscriberUri":{}}}}}]} Status:{ObservedGeneration:1 Conditions:[{Type:AddedToChannel Status:True Severity: LastTransitionTime:{Inner:2021-03-25 15:36:47 +0000 UTC} Reason: Message:} {Type:ChannelReady Status:Unknown Severity: LastTransitionTime:{Inner:2021-03-25 15:36:48 +0000 UTC} Reason:SubscriptionNotMarkedReadyByChannel Message:Failed to get subscription status: subscription "kafka-sub-1" not present in channel "kafka-sub-ready-channel" subscriber's list} {Type:Ready Status:Unknown Severity: LastTransitionTime:{Inner:2021-03-25 15:36:48 +0000 UTC} Reason:SubscriptionNotMarkedReadyByChannel Message:Failed to get subscription status: subscription "kafka-sub-1" not present in channel "kafka-sub-ready-channel" subscriber's list} {Type:ReferencesResolved Status:True Severity: LastTransitionTime:{Inner:2021-03-25 15:36:47 +0000 UTC} Reason: Message:}] Annotations:map[]}}
    operation.go:75: Failed to get &TypeMeta{Kind:Subscription,APIVersion:messaging.knative.dev/v1beta1,}-kafka-sub-1 ready: <nil>
```

lberk commented:
/retest

devguyio commented:
@lberk a generic question before reviewing the PR, does creating a second subscription manually also never gets ready?

lberk commented:
@devguyio dunno, the task was to write a repeatable, e2e test to ensure the subscription is only ready when it is actually ready.  Is something done incorrectly in the e2e test? it repeatedly fails for me.  Does it work for you?

matzew commented:
If it just works manually... we have a problem :smile:

devguyio commented:
@lberk @matzew I needed to understand more context cause I'm on my phone for the rest of the day, answering that wouldn't helped but I understand. I'll look at this on Monday. Sorry folks.

devguyio commented:
@lberk so it does work manually, and that means the test is exposing a corner case that doesn't show up when there's a delay between the steps in the test (i.e. when manually executing the same steps of the test). I believe that corner case manifests itself when the subscription is created before the pod informers for the dispatcher pods are synced. I confirmed this when putting a sleep before creating sub 1 and after scaling the dispatcher pod made the test pass. 

I'll work on a fix for this corner case.

devguyio commented:
@lberk #487 should fix it.

lberk committed:
run hack/update-codegen.sh

devguyio commented:
/retest pull-knative-sandbox-eventing-kafka-build-tests

devguyio commented:
/test pull-knative-sandbox-eventing-kafka-build-tests

devguyio commented:
@lberk I'm checking out what's wrong

devguyio commented:
/retest

lberk committed:
Modify build constraints for test

devguyio commented:
/test all

devguyio commented:
@lberk can you rebase on main?

matzew commented:
Can we rebase it to the latest ? 

sounds like this exposes the `SubscriptionNotMarkedReadyByChannel` problem

lberk commented:
I rebased it 4 days ago, can do so again if folks are actually using this to debug the underlying issue and if we're looking to get this merged.  Otherwise it might be worth to wait rebasing if we're not going to actually work to merge it

matzew commented:
Ah!

ok. thanks for letting me know!

On Tue 13. Apr 2021 at 20:14, Lukas Berk ***@***.***> wrote:

> I rebased it 4 days ago, can do so again if folks are actually using this
> to debug the underlying issue and if we're looking to get this merged.
> Otherwise it might be worth to wait rebasing if we're not going to actually
> work to merge it
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/knative-sandbox/eventing-kafka/pull/476#issuecomment-818951470>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABGPTTJVAYCQCOBX47Q5ALTISCU3ANCNFSM4ZVZE7YA>
> .
>
-- 
Sent from Gmail Mobile

matzew commented:
@lberk looks like here is some related code for reproducing the `subscriptionnotmarkedreadybychannel` problem....

https://github.com/maschmid/kafkachannel-backed-brokers-subscriptionnotmarkedreadybychannel

matzew commented:
/retest

lberk committed:
Remove v1alpha1 references as requested

lberk committed:
Fix old reference

lberk committed:
Change subscription creation function reference

matzew commented:
@devguyio is the prober working w/ TLS/SALS enabled? Or is that not relevant at all ? :thinking:

lberk commented:
/retest

lberk commented:
/retest

devguyio commented:
@matzew it's not relevant cause there's no direct communication with Kafka.

devguyio reviewed (approved):
/lgtm

devguyio commented:
/unhold

lberk commented:
/retest

lberk commented:
/retest

lberk commented:
/test pull-knative-sandbox-eventing-kafka-integration-test-channel-consolidated
